chronograf.go

```diff
@@ -136,204 +125,25 @@ type Range struct {
-	}
-
-	// reposition to duration literal
-	durFragment := fragment[start+len(prefix):]
-
-	// init counters
-	pos := 0
-
-	// locate end of duration literal
-	for pos < len(durFragment) {
-		rn, _ := utf8.DecodeRuneInString(durFragment[pos:])
-		if unicode.IsSpace(rn) {
-			break
-		}
-		pos++
-	}
-
-	// attempt to parse what we suspect is a duration literal
-	dur, err := influxql.ParseDuration(durFragment[:pos])
-	if err != nil {
-		return dur, err
-	}
-
-	return dur, nil
-}
-
-// parseAbsolute will determine the duration between two absolute timestamps
-// found within an InfluxQL fragment following the "where" keyword. For
-// example, the fragement "time > '1985-10-25T00:01:21-0800 and time <
-// '1985-10-25T00:01:22-0800'" would yield a duration of 1m'
-func (g *GroupByVar) parseAbsolute(fragment string) (time.Duration, error) {
-	timePtn := `time\s[>|<]\s'([0-9\-T\:\.Z]+)'` // Playground: http://gobular.com/x/208f66bd-1889-4269-ab47-1efdfeeb63f0
-	re, err := regexp.Compile(timePtn)
-	if err != nil {
-		// this is a developer error and should complain loudly
-		panic("Bad Regex: err:" + err.Error())
-	}
-
-	if !re.Match([]byte(fragment)) {
-		return time.Duration(0), errors.New("absolute duration not found")
-	}
-
-	// extract at most two times
-	matches := re.FindAll([]byte(fragment), 2)
-
-	// parse out absolute times
-	durs := make([]time.Time, 0, 2)
-	for _, match := range matches {
-		durStr := re.FindSubmatch(match)
-		if tm, err := time.Parse(time.RFC3339Nano, string(durStr[1])); err == nil {
-			durs = append(durs, tm)
-		}
-	}
-
-	if len(durs) == 1 {
-		durs = append(durs, time.Now())
-	}
-
-	// reject more than 2 times found
-	if len(durs) != 2 {
-		return time.Duration(0), errors.New("must provide exactly two absolute times")
-	}
-
-	dur := durs[1].Sub(durs[0])
-
-	return dur, nil
-}
-
-func (g *GroupByVar) String() string {
-	// The function is: ((total_seconds * millisecond_converstion) / group_by) = pixels / 3
-	// Number of points given the pixels
-	pixels := float64(g.Resolution) / 3.0
-	msPerPixel := float64(g.Duration/time.Millisecond) / pixels
-	secPerPixel := float64(g.Duration/time.Second) / pixels
-	if secPerPixel < 1.0 {
-		if msPerPixel < 1.0 {
-			msPerPixel = 1.0
-		}
-		return "time(" + strconv.FormatInt(int64(msPerPixel), 10) + "ms)"
-	}
-	// If groupby is more than 1 second round to the second
-	return "time(" + strconv.FormatInt(int64(secPerPixel), 10) + "s)"
-}
-
-func (g *GroupByVar) Name() string {
-	return g.Var
-}
-
-func (g *GroupByVar) Precedence() uint {
-	return 1
+type TemplateVar struct {
+	Var    string          `json:"tempVar"` // Var is the string to replace within InfluxQL
+	Values []TemplateValue `json:"values"`  // Values are the replacement values within InfluxQL
 }
 
 // TemplateID is the unique ID used to identify a template
 type TemplateID string
 
 // Template represents a series of choices to replace TemplateVars within InfluxQL
 type Template struct {
-	BasicTemplateVar
+	TemplateVar
```

Comment:
Is this actually a good practice embedding structs like this? I can't recall specifically why, but I have the general impression that embedding structs is an antipattern. The one argument I can see is that it has a bit of indirection in that by looking at the struct you don't immediately know which attributes it exposes as json.